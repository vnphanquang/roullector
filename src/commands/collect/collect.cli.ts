import {
  Command,
  InvalidArgumentError,
  Option,
} from 'commander';

import { collect } from '$commands/collect/collect';
import {
  keyTransformCliToFunc,
  defaultCollectOptions,
  defaultKeyTransformCli,
} from '$commands/collect/collect.constants';
import type {
  CollectOptions,
  KeyTransformFn,
} from '$commands/collect/collect.types';
import { KeyTransformCli } from '$commands/collect/collect.types';
import {
  commaSeparatedList,
  number,
} from '$utils/parser';

export function collectCli() {
  return new Command('collect')
    .description('collect route data')
    .option(
      '-i, --inDir <path>',
      'directory path to collect',
      defaultCollectOptions.inDir
    )
    .option(
      '-e, --extensions <list>',
      'comma-separated list of file extensions to collect',
      commaSeparatedList,
      defaultCollectOptions.extensions,
    )
    .option(
      '-x, --ignorePatterns <list>',
      'comma-separated list of filename patterns (string literal or regex) to ignore',
      commaSeparatedList,
      defaultCollectOptions.ignorePatterns,
    )
    .option(
      '--no-output',
      'don\'t write to disk but print to console instead',
      defaultCollectOptions.output,
    )
    .option(
      '-o, --outDir <path>',
      'outputs generated files to this folder',
      defaultCollectOptions.outDir,
    )
    .option(
      '-d, --depth <number>',
      'depth of inDir to collect',
      number,
      defaultCollectOptions.depth,
    )
    .addOption(
      new Option( '-t, --keyTransform <variant>', 'how to transform route key')
        .default([])
        .argParser<KeyTransformFn[]>((value, previous) => {
          if (!(value in KeyTransformCli)) {
            throw new InvalidArgumentError(`Must be one of: ${Object.keys(KeyTransformCli).join(', ')}`);
          }
          const valued = keyTransformCliToFunc[KeyTransformCli[value] as KeyTransformCli];
          return [...previous, valued];
        }),
    )
    .option(
      '-k, --dirkey <name>',
      'how to transform route key',
      defaultCollectOptions.dirkey,
    )
    .option(
      '--no-utils',
      'don\'t generate utils for building path with arguments',
      defaultCollectOptions.utils,
    )
    .option(
      '--no-typescript',
      'outputs files in javascript instead of typescript',
      defaultCollectOptions.typescript,
    )
    .option(
      '--verbose',
      'prints more info during operation',
      defaultCollectOptions.verbose,
    )
    .action((options: CollectOptions) => {
      if (options.keyTransform.length === 0) {
        options.keyTransform = defaultKeyTransformCli.map((k) => keyTransformCliToFunc[k]);
      }
      if (options.verbose) {
        console.info('Running "collect" with options:');
        console.table(Object.fromEntries(
          Object.entries(options).map(
            ([key, value]) => ([
              key,
              { value },
            ])
          )),
        );
      }

      const output = collect(options);
      if (!options.output) {
        console.log(output);
      }
    });
}
