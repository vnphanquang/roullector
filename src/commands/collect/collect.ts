import {
  readdirSync,
  mkdirSync,
  existsSync,
  rmdirSync,
  writeFileSync,
} from 'fs';
import { resolve } from 'path';

import type {
  CollectOptions,
  CollectOutput,
} from '$commands/collect/collect.types';
import { generateMapping } from '$commands/collect/generators/mapping';
import { generateRouteUtil } from '$commands/collect/generators/route';

import {
  defaultCollectOptions,
  MAPPING_FILENAME,
} from './collect.constants';

export function collect(options: Partial<CollectOptions> = {}): CollectOutput {
  const mergedOptions: CollectOptions = {
    ...defaultCollectOptions,
    ...options,
  };
  const { outDir, output, typescript } = options;
  const result = {
    json: null,
    route: null,
  };
  try {
    const { map } = generateMapping(mergedOptions);

    let json: CollectOutput['json'] = JSON.stringify(map, null, 2);
    let route: CollectOutput['route'] = mergedOptions ? generateRouteUtil(mergedOptions) : null;

    if (output) {
      if (!existsSync(outDir)) {
        mkdirSync(outDir, { recursive: true });
      }

      // write json
      let file = resolve(outDir, MAPPING_FILENAME);
      writeFileSync(file, json, { encoding: 'utf-8' });
      json = file;

      // write index
      const ext = typescript ? '.ts' : '.js';
      file = resolve(outDir, 'index' + ext);
      writeFileSync(file, route, { encoding: 'utf-8' });
      route = file;
    }

    result.json = json;
    result.route = route;
  } catch (error) {
    console.error(error);
    if (existsSync(outDir) && readdirSync(outDir).length === 0) {
      rmdirSync(outDir, { recursive: true });
    }
    throw error;
  }
  return result;
}
