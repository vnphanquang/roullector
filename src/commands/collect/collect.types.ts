export type CollectOptions = {
  /** input directory path to collect route data from */
  inDir: string;
  /** file extensions */
  extensions: string[];
  /** patterns to ignore files */
  ignorePatterns: (string|RegExp)[];
  /** whether to write to disk. If false will print to console in command-line mode, or returned in lib usage */
  output: boolean;
  /** route data output directory path */
  outDir: string;
  /** prints more info during operation */
  verbose: boolean;
  /** how to transform route key */
  keyTransform: KeyTransformFn[];
  /** depth of inDir to collect */
  depth: number;
  /** key to save path for directories with no index file */
  dirkey: string;
  /** generate utils for building path with arguments */
  utils: boolean;
  /** whether to output files in typescript */
  typescript: boolean;
};

export type CollectOutput = {
  /** path to generated json, if options.output is false, this is the generated output itself */
  json: null|string;
  /** path to generated route util, if options.output is false, this is the generated output itself */
  route: null|string;
};

export enum KeyTransformCli {
  none = 'none',
  camelCase = 'camelCase',
  dollarArg = 'dollarArg',
}

export type KeyTransformFn = (key: string, original: string) => string;
