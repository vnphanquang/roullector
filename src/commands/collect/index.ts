export * from './collect';
export * from './collect.types';
export { defaultCollectOptions } from '$commands/collect/collect.constants';
