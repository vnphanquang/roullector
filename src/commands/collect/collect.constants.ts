import type {
  CollectOptions,
  KeyTransformFn
} from '$commands/collect/collect.types';
import {KeyTransformCli} from '$commands/collect/collect.types';
import {
  camelCasify,
  dollarArgify,
} from '$utils/transformer';

export const MAPPING_FILENAME = 'routes.json';
export const UTIL_ROUTE_FILENAME = 'route';

export const defaultCollectOptions: CollectOptions = {
  inDir: 'src/routes',
  extensions: ['.svelte'],
  ignorePatterns: [/^_/],
  output: true,
  outDir: 'src/generated/routing',
  typescript: true,
  keyTransform: [dollarArgify, camelCasify],
  verbose: false,
  depth: Infinity,
  dirkey: '__dir',
  utils: true,
};

export const defaultKeyTransformCli: KeyTransformCli[] = [
  KeyTransformCli.camelCase,
  KeyTransformCli.dollarArg,
];

export const keyTransformCliToFunc: Record<KeyTransformCli, KeyTransformFn> = {
  none: (str) => str,
  camelCase: camelCasify,
  dollarArg: dollarArgify,
};
