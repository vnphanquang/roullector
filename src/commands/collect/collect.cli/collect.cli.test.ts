import { resolve } from 'path';

import {
  vol,
  fs as memfs,
} from 'memfs';

import { collectCli } from '$commands/collect/collect.cli';
import {
  defaultCollectOptions,
  MAPPING_FILENAME,
} from '$commands/collect/collect.constants';
import { FILES } from '$tests/fixtures/memfs';

jest.mock('fs', () => {
  const fs = jest.requireActual('fs');
  const sourcePath = '../generators/route/route.source.ts';
  const sourceOfRouteUtil = fs.readFileSync(resolve(__dirname, sourcePath), 'utf-8');
  vol.fromJSON({
    ['./generators/route.source.ts']: sourceOfRouteUtil,
  }, resolve(__dirname, '..'));

  return memfs;
});

const savedJSON = vol.toJSON();
beforeEach(() => {
  vol.fromNestedJSON(FILES, defaultCollectOptions.inDir);
  vol.fromJSON(savedJSON, resolve(__dirname, '..'));
});
afterEach(() => {
  vol.reset();
});

test('cli collect: default => generate files exist', () => {
  const writeSpy = jest.spyOn(console, 'log').mockImplementation();
  const program = collectCli().exitOverride();
  program.parse(['node', 'collect']);
  expect(writeSpy).toHaveBeenCalledTimes(0);
  const { outDir } = defaultCollectOptions;
  const json = resolve(outDir, MAPPING_FILENAME);
  expect(memfs.existsSync(json)).toBe(true);
  const index = resolve(outDir, 'index.ts');
  expect(memfs.existsSync(index)).toBe(true);
});

test('cli collect: default --no-output no-arg', () => {
  const writeSpy = jest.spyOn(console, 'log').mockImplementation();
  const program = collectCli().exitOverride();
  program.parse(['node', 'collect', '--no-output']);
  expect(writeSpy).toHaveBeenCalledTimes(1);
  expect(writeSpy.mock.calls[writeSpy.mock.calls.length - 1][0]).toMatchSnapshot('cli collect: default');
});

test('cli collect: default --no-output with all default args', () => {
  const writeSpy = jest.spyOn(console, 'log').mockImplementation();
  const program = collectCli().exitOverride();
  program.parse([
    'node',
    'collect',
    '--no-output',
    '--inDir=src/routes',
    '--extensions=.svelte',
    '--ignorePatterns=^_',
    '--outDir=src/generated/routing',
    '--depth=Infinity',
    '--keyTransform=dollarArg',
    '--keyTransform=camelCase',
  ]);
  expect(writeSpy).toHaveBeenCalledTimes(1);
  expect(writeSpy.mock.calls[writeSpy.mock.calls.length - 1][0]).toMatchSnapshot('cli collect: default');
});

test('cli collect: verbose', () => {
  const logInfo = jest.spyOn(console, 'info').mockImplementation();
  const tableSpy = jest.spyOn(console, 'table').mockImplementation();
  const program = collectCli().exitOverride();
  program.parse(['node', 'collect', '--no-output', '--verbose']);
  expect(logInfo).toHaveBeenCalled();
  expect(tableSpy).toHaveBeenCalled();
});

test('cli collect: invalid keyTransform choice', () => {
  const program = collectCli().exitOverride();
  expect(() => {
    program.parse(['node', 'collect', '--no-output', '--keyTransform=invalid']);
  }).toThrow();
});
