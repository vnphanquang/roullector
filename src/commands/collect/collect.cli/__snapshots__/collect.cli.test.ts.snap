// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli collect: default --no-output no-arg: cli collect: default 1`] = `
Object {
  "json": "{
  \\"admin\\": {
    \\"users\\": {
      \\"$id\\": {
        \\"index\\": \\"/admin/users/[id]\\",
        \\"posts\\": {
          \\"$postId\\": \\"/admin/users/[id]/posts/[post_id]\\",
          \\"l$shortLink$L\\": \\"/admin/users/[id]/posts/l-[short-link]-l\\",
          \\"s$slug\\": \\"/admin/users/[id]/posts/s-[slug]\\",
          \\"__dir\\": \\"/admin/users/[id]/posts\\"
        }
      },
      \\"index\\": \\"/admin/users\\"
    },
    \\"__dir\\": \\"/admin\\"
  },
  \\"index\\": \\"/\\",
  \\"me\\": \\"/me\\",
  \\"signIn\\": \\"/sign-in\\",
  \\"signUp\\": \\"/sign-up\\"
}",
  "route": "export { default as AppRoutes } from './routes.json';
/**
 * build a complete path with injected arguments
 * @param path {string} based path
 * @param args {string[]} arguments to inject
 */
export function route(path: string, ...args: string[]): string {
  const params = path.match(/\\\\[[a-zA-Z_-]+\\\\]/g) ?? [];
  for (const i in params) {
    path = path.replace(params[i], args[i]);
  }
  return path;
}

",
}
`;
